presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -776,22 +788,80 @@ public Page getNextPage()
                 if (qualifyingSet == null) {
                     qualifyingSet = new QualifyingSet();
                 }
-                qualifyingSet.setRange((int) currentGroupRowCount);
+                qualifyingSet.setRange(0, Math.min(ariaBatchRows, (int) currentGroupRowCount));
                 reader.setQualifyingSets(qualifyingSet, null);
+                if ((currentRowGroup & 0x3) == 0) {
+                    // Decay row size stats every 4 groups.
+                    numAriaRows /= 2;
+                    numAriaBytes /= 2;
+                }
                 if (reorderFilters && (currentRowGroup & 0x3) != 0 && currentRowGroup != 0) {
                     // Reconsider filter order every 4 row groups.
                     reader.maybeReorderFilters();
                 }
             }
-            qualifyingSet.clearTruncationPosition();
-            reader.advance();
+            if (qualifyingSet.isEmpty()) {
+                qualifyingSet.setRange(qualifyingSet.getEnd(), Math.min(qualifyingSet.getEnd() + ariaBatchRows, (int) currentGroupRowCount));
+            }
+            int bytesBeforeAdvance = reader.getResultSizeInBytes();
+            int numResultsBeforeAdvance = reader.getNumResults();
+            try {
+                reader.advance();
+            }
+            catch (BatchTooLargeException e) {
+                // The reader ran out of budget. Retry with a smaller
+                // input qualifying set. First remove any values added
+                // by the last advance.  The input qualifying set is
+                // unaltered when the reader throws this exception.
+                reader.compactValues(new int[0], numResultsBeforeAdvance, 0);
+                verify(ariaBatchRows > MIN_BATCH_ROWS, "Running out of budget with minimum batch size");
+                ariaBatchRows = Math.max(MIN_BATCH_ROWS, ariaBatchRows / 8);
+                int[] rows = qualifyingSet.getPositions();
```

Comment:
Can this logic be simplified to a single call to qualifyingSet.setRange(qualifyingSet.getBegin(), ...)?